docs/arch/adr-001.md

```diff
@@ -0,0 +1,38 @@
+# ADR-1: Code Generation
+
+## Context
+
+Serializing and deserializing JSON in Java works best when performing it to/from POJOs.
+While it is possible to serialize into generic `JsonNode` or `Map` objects this results in untyped values that
+must be referenced via String keys, meaning the compiler is limited in what it can check for us.
+
+The classes required for JSON serialization and deserialization are reptitive and error prone to write by hand.
+[Immutables](https://immutables.github.io/) helps to some extent, but there are still many classes and fields to be written.
+
+The [Discord API documentation](https://discordapp.com/developers/docs/intro) is mostly consistent with how it documents
+the JSON structures it returns and accepts.
+
+Code generation is often difficult to use with IDEs.
+Code is often referring to classes that will not be committed, so have to be generated by the IDE.
+
+## Decision
+
+We will use code generation to generate the POJOs to be used for JSON serialization and deserialization.
+
+We will generate classes that make use of Immutables.
+
+## Consequences
+
+The risk of errors when transcribing the Discord API are reduced.
+The specifications can by copy/pasted directly from the documentation and used as the source for code generation.
+
+By using Immutables we reduce the amount of code we must generate.
+For example, Immutables will provide with us with implementations of `equals` and `hashCode`,
+along with a builder for each class.
+
+Inconsistencies in the documentation format mean some manual changes are needed after copy/paste.
+Care must be taken in managing these manual changes if (when) the Discord API is updated.
+
+Difficulties in working with an IDE should not be undersold, but Immutables already requires working with code generation.
+By limiting code generation to ErisCasper-Data we reduce the difficulties for developers working on ErisCasper.java
```

Comment:
Yes and no? There's still a good amount of real-time annotation code processing that happens in Java which needs to be addressed. Should we consider moving all of the annotation based code generations to Data?

Reply:
Isn't this addressed with the "but Immutables requires working with code generation" ?

As in, considering how this works with an IDE is a factor, but since we're already relying on annotations stuff it's not making things worse.

Reply:
Yeah, my main question was the consideration of moving all codegen from annotations to Data from .Java since there's still a lot of codegen there. imo it would make a bit more sense if Data is here to handle codegen.

Reply:
For me the question of whether all that should be in Data is outside the scope of this ADR.

I do need to clarify this paragraph though. It should probably be more like:

> Using classes created by code generation using a custom script is not supported by IDEs. By isolating this code generation to its own library we avoid introducing friction when working with ErisCasper.java. This is in contrast to the code generation performed by Immutables that does have IDE support.

Reply:
Yeah that makes a bit more sense.